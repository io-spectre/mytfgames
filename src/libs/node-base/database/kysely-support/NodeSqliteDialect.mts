/*
The MIT License (MIT)

Copyright (c) 2022 Sami KoskimÃ¤ki

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import {
  DatabaseIntrospector,
  Dialect,
  DialectAdapter,
  Driver,
  Kysely,
  QueryCompiler,
  SqliteAdapter,
  SqliteIntrospector,
  SqliteQueryCompiler,
} from "kysely";

import { NodeSqliteDialectConfig } from "./NodeSqliteConfig.mjs";
import { NodeSqliteDriver } from "./NodeSqliteDriver.mjs";

/**
 * SQLite dialect that uses the [better-sqlite3](https://github.com/JoshuaWise/better-sqlite3) library.
 *
 * The constructor takes an instance of {@link NodeSqliteDialectConfig}.
 *
 * ```ts
 * import Database from 'better-sqlite3'
 *
 * new SqliteDialect({
 *   database: new Database('db.sqlite')
 * })
 * ```
 *
 * If you want the pool to only be created once it's first used, `database`
 * can be a function:
 *
 * ```ts
 * import Database from 'better-sqlite3'
 *
 * new SqliteDialect({
 *   database: async () => new Database('db.sqlite')
 * })
 * ```
 */
export class NodeSqliteDialect implements Dialect {
  readonly #config: NodeSqliteDialectConfig;

  constructor(config: NodeSqliteDialectConfig) {
    this.#config = Object.freeze({ ...config });
  }

  createDriver(): Driver {
    return new NodeSqliteDriver(this.#config);
  }

  createQueryCompiler(): QueryCompiler {
    return new SqliteQueryCompiler();
  }

  createAdapter(): DialectAdapter {
    return new SqliteAdapter();
  }

  createIntrospector(db: Kysely<unknown>): DatabaseIntrospector {
    return new SqliteIntrospector(db);
  }
}
