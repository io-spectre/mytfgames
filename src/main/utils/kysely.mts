import { Kysely, sql, Transaction } from "kysely";

/**
 * Run migration logic in a transaction on a sqlite database.
 *
 * This function will disable foreign key checks, run the migration, and then
 * re-enable foreign key checks. This is necessary because sqlite does not offer
 * adequate ALTER TABLE support and does not support disabling foreign key
 * checks in a transaction.
 *
 * The any types reflect that the database is not in sync with the types of the
 * ORM as per the Kysely documentation.
 *
 * @param db - The database to run the migration on.
 * @param migration - The migration to run.
 */
export async function sqliteSafeMigration(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  db: Kysely<any>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  migration: (trx: Transaction<any>) => Promise<void>,
): Promise<void> {
  await sql`PRAGMA foreign_keys=0`.execute(db);
  try {
    await db.transaction().execute(migration);
  } finally {
    await sql`PRAGMA foreign_keys=1`.execute(db);
  }
}
