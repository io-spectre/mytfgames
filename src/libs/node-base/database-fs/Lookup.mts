import { sql } from "kysely";

import { AppQueryCreator, WellKnownDirectory } from "$node-base/database";
import { makeLogicError } from "$pure-base/utils";

/**
 * Resolves a path to a node number. The path is resolved relative to the root node.
 *
 * @param qc The query creator, either the database connection or a transaction
 * @param path The path to resolve
 * @returns The node number of the resolved path or undefined if not found
 */
export async function dbfsFindNodeNoByPath(
  qc: AppQueryCreator,
  path: string,
): Promise<bigint | undefined>;
/**
 * Resolves a path to a node number. The path is resolved relative to the anchor node.
 * If the path is absolute, the anchor node must be the root node.
 *
 * @param qc The query creator, either the database connection or a transaction
 * @param anchor Search starting point, defaults to the root node
 * @param path The path to resolve
 * @returns The node number of the resolved path or undefined if not found
 */
export async function dbfsFindNodeNoByPath(
  qc: AppQueryCreator,
  anchor: bigint,
  path: string,
): Promise<bigint | undefined>;
export async function dbfsFindNodeNoByPath(
  qc: AppQueryCreator,
  anchorOrPath: bigint | string,
  path?: string,
): Promise<bigint | undefined> {
  let parentNo: bigint;
  if (typeof anchorOrPath !== "string") {
    if (path === undefined) {
      throw makeLogicError("Missing path argument");
    }
    parentNo = anchorOrPath;
  } else {
    if (path !== undefined) {
      throw makeLogicError("Unexpected path argument");
    }
    parentNo = WellKnownDirectory.ROOT;
    path = anchorOrPath;
  }
  if (parentNo !== WellKnownDirectory.ROOT && path.startsWith("/")) {
    throw makeLogicError(
      "Tried to resolve an absolute path with a parent node",
    );
  }
  const pathSegments = path.split("/").filter((segment) => segment.length > 0);
  if (pathSegments.length === 0) {
    // ensure that the parent node exists
    const rr = await qc
      .selectFrom("node")
      .select("node_no")
      .where("node_no", "=", parentNo)
      .executeTakeFirst();
    return rr === undefined ? rr : BigInt(rr.node_no);
  }
  const pathAsJsonArray = JSON.stringify(pathSegments);

  const rr = await qc
    .withRecursive("path_segment", (ctec) =>
      ctec
        .selectFrom(sql`json_each(${pathAsJsonArray})`.as("pathp"))
        .select(sql<number>`row_number() over ()`.as("segment_no"))
        .select(sql<string>`pathp.value`.as("segment")),
    )
    .withRecursive("ids", (ctec) =>
      ctec
        .selectFrom(["node_member"])
        .select(["node_member.node_no", "node_member.node_no_parent"])
        .select(sql<bigint>`2`.as("i"))
        .where("node_member.node_no_parent", "=", parentNo)
        .where("node_member.name", "=", pathSegments[0]!)
        .unionAll((uac) =>
          uac
            .selectFrom("ids")
            .innerJoin("path_segment", "path_segment.segment_no", "ids.i")
            .innerJoin("node_member", (join) =>
              join
                .onRef("node_member.node_no_parent", "=", "ids.node_no")
                .onRef("node_member.name", "=", "path_segment.segment"),
            )
            .select(["node_member.node_no", "node_member.node_no_parent"])
            .select(sql<bigint>`ids.i + 1`.as("i"))
            .orderBy("i"),
        ),
    )
    .selectFrom("ids")
    .selectAll()
    .where("i", "=", BigInt(pathSegments.length) + 1n)
    .executeTakeFirst();

  return rr === undefined ? rr : BigInt(rr.node_no);
}
