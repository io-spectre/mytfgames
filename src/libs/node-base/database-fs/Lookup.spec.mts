import { DatabaseSync } from "node:sqlite";

import { Kysely, Migrator } from "kysely";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

import {
  AppDatabase,
  ViteMigrationProvider,
  WellKnownDirectory,
} from "$node-base/database";

import { NodeSqliteDialect } from "../database/kysely-support/NodeSqliteDialect.mjs";
import { dbfsFindNodeNoByPath } from "./Lookup.mjs";

describe("database-fs:Lookup", () => {
  let db: Kysely<AppDatabase>;

  beforeEach(async () => {
    db = new Kysely<AppDatabase>({
      dialect: new NodeSqliteDialect({
        database: new DatabaseSync(":memory:"),
      }),
      log: ["error"],
    });
    const migrator = new Migrator({
      db,
      provider: new ViteMigrationProvider(),
    });
    const migration = await migrator.migrateToLatest();
    if (migration.error != null) {
      throw new Error("Failed to migrate the database schema", {
        cause: migration.error,
      });
    }

    await db
      .insertInto("node")
      .values([
        { node_type: "D", node_no: 2001n },
        { node_type: "D", node_no: 2002n },
        { node_type: "D", node_no: 2003n },
        { node_type: "D", node_no: 2004n },
        { node_type: "D", node_no: 2005n },
        { node_type: "D", node_no: 2006n },
        { node_type: "D", node_no: 2007n },
      ])
      .execute();
    await db
      .insertInto("node_directory")
      .values([
        { directory_no: 2001n },
        { directory_no: 2002n },
        { directory_no: 2003n },
        { directory_no: 2004n },
        { directory_no: 2005n },
        { directory_no: 2006n },
        { directory_no: 2007n },
      ])
      .execute();
    await db
      .insertInto("node_member")
      .values([
        { node_no: 2001n, node_no_parent: WellKnownDirectory.TMP, name: "a" },
        { node_no: 2002n, node_no_parent: WellKnownDirectory.ROOT, name: "b" },
        { node_no: 2003n, node_no_parent: 2001n, name: "c" },
        { node_no: 2004n, node_no_parent: 2001n, name: "d" },
        { node_no: 2005n, node_no_parent: 2004n, name: "e" },
        { node_no: 2006n, node_no_parent: 2004n, name: "f" },
        { node_no: 2007n, node_no_parent: 2005n, name: "g" },
      ])
      .execute();
  });
  afterEach(() => {
    return db.destroy();
  });

  describe("dbfsFindNodeNoByPath", () => {
    it("should throw an error for absolute paths with a parent node", async () => {
      await expect(
        dbfsFindNodeNoByPath(db, 2005n, "/tmp/a/d/e"),
      ).rejects.toThrowErrorMatchingSnapshot();
    });
    it("should retrieve the correct node no from root", async () => {
      await expect(dbfsFindNodeNoByPath(db, "/tmp/a/d/e")).resolves.toBe(2005n);
    });
    it("should retrieve the correct node no from an intermediate directory", async () => {
      await expect(
        dbfsFindNodeNoByPath(db, WellKnownDirectory.TMP, "a/d/f"),
      ).resolves.toBe(2006n);
    });
    it("should return undefined for non-existent paths", async () => {
      await expect(
        dbfsFindNodeNoByPath(db, "/tmp/a/d/x"),
      ).resolves.toBeUndefined();
    });
    it("should return the parent node no for empty paths", async () => {
      await expect(dbfsFindNodeNoByPath(db, 2005n, "")).resolves.toBe(2005n);
    });
    it("should return undefined for empty paths with a non-existent parent", async () => {
      await expect(
        dbfsFindNodeNoByPath(db, 9999n, ""),
      ).resolves.toBeUndefined();
    });
    it("should return undefined for a path starting with a non-existent parent", async () => {
      await expect(
        dbfsFindNodeNoByPath(db, 9999n, "a/d/e"),
      ).resolves.toBeUndefined();
    });
  });
});
