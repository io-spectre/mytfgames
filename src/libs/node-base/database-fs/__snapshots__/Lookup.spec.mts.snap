// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`database-fs:Lookup > dbfsFindNodeNoByPath > should throw an error for absolute paths with a parent node 1`] = `[Error: Tried to resolve an absolute path with a parent node]`;
